packages/block-editor/src/hooks/content-lock-ui.js

```diff
@@ -0,0 +1,161 @@
+/**
+ * WordPress dependencies
+ */
+import { ToolbarButton } from '@wordpress/components';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { useDispatch, useSelect } from '@wordpress/data';
+import { addFilter } from '@wordpress/hooks';
+import { __ } from '@wordpress/i18n';
+import { useEffect, useRef, useCallback } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../store';
+import { BlockControls } from '../components';
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+function StopEditingAsBlocksOnOutsideSelect( {
+	clientId,
+	stopEditingAsBlock,
+} ) {
+	const isBlockOrDescendantSelected = useSelect(
+		( select ) => {
+			const { isBlockSelected, hasSelectedInnerBlock } =
+				select( blockEditorStore );
+			return (
+				isBlockSelected( clientId ) ||
+				hasSelectedInnerBlock( clientId, true )
+			);
+		},
+		[ clientId ]
+	);
+	useEffect( () => {
+		if ( ! isBlockOrDescendantSelected ) {
+			stopEditingAsBlock();
+		}
+	}, [ isBlockOrDescendantSelected ] );
+	return null;
+}
+
+export const withBlockControls = createHigherOrderComponent(
+	( BlockEdit ) => ( props ) => {
+		const { getBlockListSettings, getSettings } =
+			useSelect( blockEditorStore );
+		const focusModeToRevert = useRef();
+		const { templateLock, isLockedByParent, isEditingAsBlocks } = useSelect(
+			( select ) => {
+				const {
+					__unstableGetContentLockingParent,
+					getTemplateLock,
+					__unstableGetTemporarilyEditingAsBlocks,
+				} = select( blockEditorStore );
+				return {
+					templateLock: getTemplateLock( props.clientId ),
+					isLockedByParent: !! __unstableGetContentLockingParent(
+						props.clientId
+					),
+					isEditingAsBlocks:
+						__unstableGetTemporarilyEditingAsBlocks() ===
+						props.clientId,
+				};
+			},
+			[ props.clientId ]
+		);
+
+		const {
+			updateSettings,
+			updateBlockListSettings,
+			__unstableSetTemporarilyEditingAsBlocks,
+		} = useDispatch( blockEditorStore );
+		const isContentLocked =
+			! isLockedByParent && templateLock === 'noContent';
+		const {
+			__unstableMarkNextChangeAsNotPersistent,
+			updateBlockAttributes,
+		} = useDispatch( blockEditorStore );
+
+		const stopEditingAsBlock = useCallback( () => {
+			__unstableMarkNextChangeAsNotPersistent();
+			updateBlockAttributes( props.clientId, {
+				templateLock: 'noContent',
+			} );
+			updateBlockListSettings( props.clientId, {
```

Comment:
I've seen that this updates the `blockListBlockSettings` key in the block-editor store, duplicating the `templateLock` data we already have in `block.attributes`. Why is this necessary?